Project References don't resolve ambient Definitions

**TypeScript Version:**  3.3.0-dev.20181213

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:**
Project References
Composite Project
Ambient Definitions

**Code**
(minimum example, see repository link below for complete setup)
*tsconfig_base.json*
```json
{
  "compilerOptions": {
    "target": "es5",
    "module": "none",
    "declaration": true,
    "declarationMap": true,
    "sourceMap": true,
    "strict": true,
    "composite": true
  }
}
```
*utils/tsconfig.json*
```json
{
  "extends": "../tsconfig_base.json",
  "include": [
    "*.ts",
    "*.d.ts"
  ]
}
```
*utils/utils.d.ts*
```ts
declare function leftPad(s: string, n: number): string;
```
*converter/tsconfig.json*
```json
{
  "extends": "../tsconfig_base.json",
  "include": [
    "*.ts",
    "*.d.ts"
  ],
  "references": [
    { "path": "../utils" }
  ]
}
```
*converter/length.ts*
```ts
function formatInches(i: number) {
    if (i < 12) {
        return `${i}"`;
    }
    const ft = Math.floor(i / 12);
    i = i - ft * 12;
    return `${ft}' ${leftPad(i.toString(), 2)}`;
}
```

**Expected behavior:**
Running `tsc -b converter` compiles successfully.

**Actual behavior:**
```
tsc -b converter --verbose
[13:10:29] Projects in this build:
    * utils/tsconfig.json
    * converter/tsconfig.json

[13:10:29] Project 'utils/tsconfig.json' is up to date because newest input 'utils/moreUtils.ts' is older than oldest output 'utils/moreUtils.js.map'

[13:10:29] Project 'converter/tsconfig.json' is out of date because output file 'converter/length.js' does not exist

[13:10:29] Building project '[...]/converter/tsconfig.json'...

converter/length.ts:18:22 - error TS2304: Cannot find name 'leftPad'.

18     return `${ft}' ${leftPad(i.toString(), 2)}`;
                        ~~~~~~~


Found 1 error.
```
(Local paths redacted)

As a side note, when compiling the packages into outFiles the compilation works just fine.
(This can be seen in @RyanCavanaugh's example repository from which mine is forked / when reverting to commit 0f66d20 of my example repository.)

~~**Playground Link:**~~

**Example Repository:**
https://github.com/LucaVazz/ts-project-references-for-ambient

**Related Issues:** #22997 #25600 #26595 #26689 #26867

--- a/src/compiler/program.ts
+++ b/src/compiler/program.ts
@@ -825,11 +825,16 @@ namespace ts {
                 }
                 if (rootNames.length) {
                     for (const parsedRef of resolvedProjectReferences) {
-                        if (parsedRef) {
-                            const out = parsedRef.commandLine.options.outFile || parsedRef.commandLine.options.out;
-                            if (out) {
-                                const dtsOutfile = changeExtension(out, ".d.ts");
-                                processSourceFile(dtsOutfile, /*isDefaultLib*/ false, /*ignoreNoDefaultLib*/ false, /*packageId*/ undefined);
+                        if (!parsedRef) continue;
+                        const out = parsedRef.commandLine.options.outFile || parsedRef.commandLine.options.out;
+                        if (out) {
+                            processSourceFile(changeExtension(out, ".d.ts"), /*isDefaultLib*/ false, /*ignoreNoDefaultLib*/ false, /*packageId*/ undefined);
+                        }
+                        else if (getEmitModuleKind(parsedRef.commandLine.options) === ModuleKind.None) {
+                            for (const fileName of parsedRef.commandLine.fileNames) {
+                                if (!fileExtensionIs(fileName, Extension.Dts) && hasTSFileExtension(fileName)) {
+                                    processSourceFile(getOutputDeclarationFileName(fileName, parsedRef.commandLine), /*isDefaultLib*/ false, /*ignoreNoDefaultLib*/ false, /*packageId*/ undefined);
+                                }
                             }
                         }
                     }

--- a/src/testRunner/unittests/tsbuild/outFile.ts
+++ b/src/testRunner/unittests/tsbuild/outFile.ts
@@ -449,7 +449,6 @@ namespace ts {
             changeCompilerVersion(host);
             builder.buildAllProjects();
             host.assertDiagnosticMessages(
-                // TODO:: This should build all instead
                 getExpectedDiagnosticForProjectsInBuild(relSources[project.first][source.config], relSources[project.second][source.config], relSources[project.third][source.config]),
                 [Diagnostics.Project_0_is_out_of_date_because_output_for_it_was_generated_with_version_1_that_differs_with_current_version_2, relSources[project.first][source.config], fakes.version, version],
                 [Diagnostics.Building_project_0, sources[project.first][source.config]],
@@ -855,6 +854,48 @@ ${internal} enum internalEnum { a, b, c }`);
                     baselineOnly: true
                 });
             });
+        });
+
+        it("non module projects without prepend", () => {
+            const fs = outFileFs.shadow();
+            // No prepend
+            replaceText(fs, sources[project.third][source.config], `{ "path": "../first", "prepend": true }`, `{ "path": "../first" }`);
+            replaceText(fs, sources[project.third][source.config], `{ "path": "../second", "prepend": true }`, `{ "path": "../second" }`);
+
+            // Non Modules
+            replaceText(fs, sources[project.first][source.config], `"composite": true,`, `"composite": true, "module": "none",`);
+            replaceText(fs, sources[project.second][source.config], `"composite": true,`, `"composite": true, "module": "none",`);
+            replaceText(fs, sources[project.third][source.config], `"composite": true,`, `"composite": true, "module": "none",`);
+
+            // Own file emit
+            replaceText(fs, sources[project.first][source.config], `"outFile": "./bin/first-output.js",`, "");
+            replaceText(fs, sources[project.second][source.config], `"outFile": "../2/second-output.js",`, "");
+            replaceText(fs, sources[project.third][source.config], `"outFile": "./thirdjs/output/third-output.js",`, "");
+
+            const host = new fakes.SolutionBuilderHost(fs);
+            const builder = createSolutionBuilder(host);
+            builder.buildAllProjects();
+            host.assertDiagnosticMessages(
+                getExpectedDiagnosticForProjectsInBuild(relSources[project.first][source.config], relSources[project.second][source.config], relSources[project.third][source.config]),
+                [Diagnostics.Project_0_is_out_of_date_because_output_file_1_does_not_exist, relSources[project.first][source.config], "src/first/first_PART1.js"],
+                [Diagnostics.Building_project_0, sources[project.first][source.config]],
+                [Diagnostics.Project_0_is_out_of_date_because_output_file_1_does_not_exist, relSources[project.second][source.config], "src/second/second_part1.js"],
+                [Diagnostics.Building_project_0, sources[project.second][source.config]],
+                [Diagnostics.Project_0_is_out_of_date_because_output_file_1_does_not_exist, relSources[project.third][source.config], "src/third/third_part1.js"],
+                [Diagnostics.Building_project_0, sources[project.third][source.config]]
+            );
+            const expectedOutputFiles = flatMap(sources, ([config, ts]) => [
+                removeFileExtension(config) + Extension.TsBuildInfo,
+                ...flatMap(ts, f => [
+                    removeFileExtension(f) + Extension.Js,
+                    removeFileExtension(f) + Extension.Js + ".map",
+                    removeFileExtension(f) + Extension.Dts,
+                    removeFileExtension(f) + Extension.Dts + ".map",
+                ])
+            ]);
+            for (const output of expectedOutputFiles) {
+                assert(fs.existsSync(output), `Expect file ${output} to exist`);
+            }
         });
     });
 }

